src/transforms/aggregate.rs

```diff
@@ -0,0 +1,420 @@
+use crate::{
+    internal_events::{AggregateEventRecorded, AggregateFlushed},
+    transforms::{
+        TaskTransform,
+        Transform,
+    },
+    config::{DataType, GlobalOptions, TransformConfig, TransformDescription},
+    event::{
+        metric,
+        Event,
+        EventMetadata,
+    },
+};
+use async_stream::stream;
+use futures::{stream, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    time::{Duration},
+};
+
+#[derive(Deserialize, Serialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct AggregateConfig {
+    /// The interval between flushes in milliseconds.
+    pub interval_ms: Option<u64>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<AggregateConfig>("aggregate")
+}
+
+impl_generate_config_from_default!(AggregateConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "aggregate")]
+impl TransformConfig for AggregateConfig {
+    async fn build(&self, _globals: &GlobalOptions) -> crate::Result<Transform> {
+        Aggregate::new(self).map(Transform::task)
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "aggregate"
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Aggregate {
+    interval: Duration,
+    map: HashMap<metric::MetricSeries, metric::MetricData>,
+}
+
+impl Aggregate {
+    pub fn new(config: &AggregateConfig) -> crate::Result<Self> {
+        Ok(Self {
+            interval: Duration::from_millis(config.interval_ms.unwrap_or(10 * 1000)),
+            map: HashMap::new(),
+        })
+    }
+
+    fn record(&mut self, event: Event) {
+        let metric = event.as_metric();
+        let series = metric.series();
+        let data = metric.data();
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                match self.map.get_mut(&series) {
+                    // We already have something, add to it, will update timestamp as well.
+                    Some(existing) => existing.update(data),
```

Comment:
I chose not to explicitly handle it since there was no meaningful thing to do in that case. Only thing I can think of to do would be to emit an internal metric like `AggregateIncompatibleValue` or something of that sort.